BoogieLang/Ast_Cfg_Transformation.thy

```diff
@@ -0,0 +1,2280 @@
+    by (simp add: endblock_skip2)
+  thus ?thesis
+  proof cases
+    assume "(valid_configuration A \<Lambda> \<Gamma> \<Omega> posts reached_bb reached_cont reached_state)" thus ?thesis by simp
+  next 
+    assume "\<not> ((valid_configuration A \<Lambda> \<Gamma> \<Omega> posts reached_bb reached_cont reached_state))"
+    hence "(\<exists> l2. (A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bigblock_next, cont0, Normal ns1'') -n\<longrightarrow>^l2 (reached_bb, reached_cont, reached_state)) \<and> (j' = Suc (Suc l2)) )" 
+      using disj_a by blast
+    thus ?thesis
+    proof -
+      obtain l2_conc where conc_trace: "(A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bigblock_next, cont0, Normal ns1'') -n\<longrightarrow>^l2_conc (reached_bb, reached_cont, reached_state))" and 
+                           succ_rel: "(j' = Suc (Suc l2_conc))"
+        using \<open>\<exists>l2. (A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bigblock_next, cont0, Normal ns1'') -n\<longrightarrow>^l2 (reached_bb, reached_cont, reached_state)) \<and> j' = Suc (Suc l2)\<close> by blast
+      show ?thesis
+        apply (rule assms(7))
+           apply (rule succ_rel)
+          apply (rule conc_trace)
+         apply (rule assms(4))
+          apply (simp)
+         apply (rule assms(5))
+        apply simp+
+        using assms(6) false_equals_not_true 
+        by blast
+    qed
+  qed
+qed
+
+text \<open>Pair 3: The starting configuration represents a point in the program before a loop and, more specifically, before the loop has been 'unwrapped'.
+              The 'wrapper' construct exists to accomodate the handling of breaks, which this theory doesn't currently cover.\<close>
+lemma wrapper_to_endblock:
+  assumes "A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb0, cont0, Normal ns) -n\<longrightarrow>^l (reached_bb, reached_cont, reached_state)"
+      and "bb0 = BigBlock name [] (Some (WhileWrapper loop)) None"
+  shows "(valid_configuration A \<Lambda> \<Gamma> \<Omega> posts reached_bb reached_cont reached_state) \<or> 
+         (\<exists> l1. (A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>((BigBlock name [] (Some loop) None), KEndBlock cont0, Normal ns) -n\<longrightarrow>^l1 (reached_bb, reached_cont, reached_state)) \<and> (l = Suc l1))"
+  using assms
+proof (cases l)
+  case 0
+  hence "(reached_bb, reached_cont, reached_state) = (bb0, cont0, Normal ns)" using assms(1) by simp
+  then show ?thesis by (simp add: Ast.valid_configuration_def assms(2))
+next
+  case 1: (Suc l1)
+  then show ?thesis 
+  proof -
+    from 1 assms obtain inter_bb inter_cont inter_state where
+      step1: "(red_bigblock A M \<Lambda> \<Gamma> \<Omega> T (bb0, cont0, Normal ns) (inter_bb, inter_cont, inter_state))" and 
+      rest: "A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(inter_bb, inter_cont, inter_state) -n\<longrightarrow>^l1 (reached_bb, reached_cont, reached_state)" 
+      by (metis (no_types, opaque_lifting) prod_cases3 relpowp_Suc_D2)
```

Comment:
replace by "by (metis (no_types) etc...)", otherwise the proofs fail on Windows.